### ISSUE TITLE
Shortcomings of BN's handling of float/vector values in the Windows x64 calling convention

### ISSUE THREAD

**Issue by alexrp** [Vector35/binaryninja-api#3191]
**Version and Platform (required):**
- Binary Ninja Version: 3.1.3504-dev
- OS: Windows
- OS Version: 10.0.25131.1000
- CPU Architecture: x64

**Bug Description:**
Consider this function:

![image](https://user-images.githubusercontent.com/44076/173206463-5d24fdc5-abe4-41c3-a184-50c18ebecd50.png)

I can glean from the disassembly that the function is supposed to have a 2nd parameter. I happen to know that it's a `float` and not a vector, but that's (probably?) impossible for BN to figure out just based on the `movaps` instructions here.

Still, I wonder if BN couldn't do slightly better: The fact that `xmm1` is read before writing anything to it strongly suggests that there at least *is* a 2nd parameter. BN could perhaps have created it and tentatively typed it as `int128_t @ xmm1` or something?

Anyhow, defining the `float delta` parameter and switching to pseudo-C, I get:

![image](https://user-images.githubusercontent.com/44076/173206812-81da354c-d7ae-42f7-9503-ce6ede26eb1f.png)

This is weird considering `FMalloc__vtable::Tick` is declared as:

```c
struct FMalloc__vtable __packed
{
    // ...
    void (* Tick)(struct FMalloc* `this`, float delta);
    // ...
};
```

The only way I could get BN to pass the 2nd parameter correctly to the virtual function was by changing the `Tick` function pointer to pass it as `float delta @ xmm1`:

![image](https://user-images.githubusercontent.com/44076/173207242-01daea3f-5080-449c-b77d-8f95227cdde3.png)

I don't think that should be necessary since the calling convention clearly dictates that it should be passed in `xmm1` anyway.

**Comment by D0ntPanic** [comment]
The calling convention API is not currently robust enough to properly decide the differences in register allocation between integer and floating point registers. These vary by platform. This will need additional APIs in the calling conventions, as well as core changes to use these APIs during automatic calling convention detection.

**Comment by alexrp** [comment]
Just wanted to check: Is this likely to be addressed in the 3.3 timeframe? (It's a fairly major productivity impediment in reverse engineering certain FP/SSE-heavy areas of multimedia applications.)

**Comment by D0ntPanic** [comment]
Progress on this issue is blocked on first resolving #4027 as the xmm vs. zmm confusion prevents any fixes from having any effect.

**Comment by alexrp** [comment]
Sorry to keep coming back to this one, but it's by far the most impactful BN issue for me. I've recently started reversing some game code that deals with complex geometric shapes, collision detection, pathfinding, etc. This is of course very dense with SSE code.

Due to this issue (and the linked related issues), the decompilation for such functions borders on entirely unusable in BN. The only case where I can *sometimes* sort of make out what's going on is when a function is operating on float values out of larger structures only; if the values are passed as scalars/vectors in XMM registers, though, no such luck.

Not asking for estimates here, I'm just curious if addressing this is on the horizon / considered a priority? FWIW, from my perspective, this is actually the only remaining blocker for BN to supplant IDA in my RE toolkit; I only use IDA's decompiler for SSE-dense functions at this stage. 🤔

**Comment by CouleeApps** [comment]
Addressing this is still a priority. While originally slated for the Coruscant release (3.5), this has proven to be significantly more involved than initially planned and will take a while to implement. We've scheduled it for the next release (Dorsai, TBD if v3.6 or v4.0), so hopefully progress should resume after we get the current release out. From the issue updating logs, you can see the rough timeline of this process. It has turned into the proverbial "issue hydra": try to fix one bug and 3 emerge in its place. All that being said, we do have solid ideas for how we are going to address this, but it may take a while.

**Comment by D0ntPanic** [comment]
Fixed in 3.6.4705
